tasks/common.py

```diff
@@ -34,7 +34,18 @@ def get_example(self, index):
 
     def __len__(self):
         start = self.start
-        stop = self.num_examples() if self.stop is None else self.stop
+        if self.stop is not None:
+            num_ex = self.num_examples()
+            stop = min(self.stop, num_ex)  # Gracefully cap at dataset size
+            if self.stop > num_ex:
+                import warnings
+                warnings.warn(
+                    f"Stop parameter ({self.stop}) exceeds dataset size ({num_ex}). "
+                    f"Using {num_ex} examples instead.",
+                    UserWarning
+                )
+        else:
+            stop = self.num_examples()
```

Comment:
I see what you mean about the constructor.

We do want to avoid emitting this warning every single time when `len()` is called though. 

What if we do something like this
```suggestion
        num_ex = self.num_examples()
        if self.stop is not None and self.stop > num_ex:
            self.stop = num_ex
            import warnings
                warnings.warn(
                    f"Stop parameter ({self.stop}) exceeds dataset size ({num_ex}). "
                    f"Using {num_ex} examples instead.",
                    UserWarning
                )
        stop = num_ex if self.stop is None else self.stop    
```
This caps `self.stop` once and emits the warning once, and from then on when this task is used, this issue shouldn't occur anymore.

I'd also move the `warnings` import to the top of the file, or better yet, use a `logger = logging.getLogger()` as the rest of the code base does.